consensus/src/quorum_store/proof_queue.rs

```diff
@@ -0,0 +1,450 @@
+        (remaining_txns_without_duplicates, self.remaining_proofs)
+    }
+
+    // Mark in the hashmap committed PoS, but keep them until they expire
+    fn mark_committed(&mut self, batches: Vec<BatchInfo>) {
+        for batch in &batches {
+            let batch_key = BatchKey::from_info(batch);
+            if let Some(Some((proof, insertion_time))) = self.batch_to_proof.get(&batch_key) {
+                counters::pos_to_commit(
+                    proof.gas_bucket_start(),
+                    insertion_time.elapsed().as_secs_f64(),
+                );
+                self.dec_remaining(&batch.author(), batch.num_txns());
+            }
+            self.batch_to_proof.insert(batch_key.clone(), None);
+            self.batches_with_txn_summary.remove(&batch_key);
+            self.txn_summary_to_batches.retain(|_, batches| {
+                batches.remove(&batch_key);
+                !batches.is_empty()
+            });
+        }
+    }
+
+    pub async fn start(mut self, mut command_rx: tokio::sync::mpsc::Receiver<ProofQueueCommand>) {
+        loop {
+            let _timer = counters::PROOF_MANAGER_MAIN_LOOP.start_timer();
+            if let Some(msg) = command_rx.recv().await {
+                match msg {
+                    ProofQueueCommand::AddProofs(proofs, response_sender) => {
+                        for proof in proofs {
+                            self.push(proof);
+                        }
+                        if let Err(e) = response_sender.send(self.remaining_txns_and_proofs()) {
+                            warn!("Failed to send response to AddProofs: {:?}", e);
+                        }
+                    },
+                    ProofQueueCommand::PullProofs {
+                        excluded_batches,
+                        max_txns,
+                        max_bytes,
+                        return_non_full,
+                        response_sender,
+                    } => {
+                        let (proofs, full) = self.pull_proofs(
+                            &excluded_batches,
+                            max_txns,
+                            max_bytes,
+                            return_non_full,
+                        );
+                        if let Err(e) = response_sender.send((proofs, full)) {
+                            warn!("Failed to send response to PullProofs: {:?}", e);
+                        }
+                    },
+                    ProofQueueCommand::MarkCommitted(batches, block_timestamp, response_sender) => {
+                        self.mark_committed(batches);
+                        self.handle_updated_block_timestamp(block_timestamp);
+                        if let Err(e) = response_sender.send(self.remaining_txns_and_proofs()) {
+                            error!("Failed to send response to MarkCommitted: {:?}", e);
+                        }
+                    },
+                    ProofQueueCommand::AddBatches(batch_summaries) => {
```

Comment:
@vusirikala given the code has been moved it's hard to tell what has changed. Are there any changes outside of `AddBatches` related code?

Reply:
In addition the previous proof_queue, I'm storing `txn_summary_to_batches` which store the list of batches for each (txn_sender, sequence_number) pair. `batches_with_txn_summary` stores the list of batchkeys for which we've received the txn summaries. I'm updating these datastructures when adding batches and handling commits. I'm using the new `remaining_txns` function when calculating the number of remaining txns.

I've added a lot of counters for testing. Will clean up the code when I have enough confidence before landing.